### PR TITLE
Fix shipment issues and optimise shipment modules

### PR DESCRIPTION
**Pull request by priyank-gupta**
Fixes 2 shipment issues, and optimises shipment modules like Coordinator, Prioritizer, Packer etc..

**Issue 1**
- Scenario: A product has 2 stock items for 2 stock locations. Each stock item has 1 quantity on hand and none is backorderable. User adds 2 quantity of that item in cart and moves to delivery step. Sees a shipment and an unshippable unit.
- Expected: User should see 2 shipments (no unshippable unit) each having 1 associated inventory unit.

**Issue 2**
- In `order#create_proposed_shipments`, inventory units are destroyed using shipments `has_many inventory_units, dependent: :delete_all` association but units which are not associated to any shipment (like unshippable units) are never destroyed.

### CODE REVIEW

File: core/app/models/spree/stock/package.rb [hunk 1]
```diff
@@ -11,7 +11,8 @@ def initialize(stock_location, contents=[])
       end
 
       def add(inventory_unit, state = :on_hand)
-        contents << ContentItem.new(inventory_unit, state) unless find_item(inventory_unit)
+        # Remove find_item check as already taken care by prioritizer
```
Comment (priyank-gupta) [inline core/app/models/spree/stock/package.rb]: Removed find_item(inventory_unit) as this scenario is already handled by Prioritizer.

**Comment by priyank-gupta** [comment]
Few Benchmarks,
OLD
- With 10 quantity, 5 on_hand, 5 backordered, same stock location
  - `#<Benchmark::Tms:0x007f97751f86f0 @label="", @real=0.13010611399658956, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.10999999999999943, @total=0.10999999999999943>`
- With 50 quantity, 25 on_hand, 25 backordered, same stock location
   - `#<Benchmark::Tms:0x007f9775964ce0 @label="", @real=0.33323719300096855, @cstime=0.0, @cutime=0.0, @stime=0.019999999999999574, @utime=0.28999999999999915, @total=0.3099999999999987>`
- With 100 quantity, 50 on_hand, 50 backordered, same stock location
   - `#<Benchmark::Tms:0x007f976210c330 @label="", @real=0.6519398770033149, @cstime=0.0, @cutime=0.0, @stime=0.019999999999999574, @utime=0.5799999999999983, @total=0.5999999999999979>`
- With 500 quantity, 250 on_hand, 250 backordered, same stock location
   - `#<Benchmark::Tms:0x007f9774ce8c50 @label="", @real=5.668459588006954, @cstime=0.0, @cutime=0.0, @stime=0.09999999999999787, @utime=5.259999999999998, @total=5.359999999999996>`
- With 1000 quantity, 500 on_hand, 500 backordered, same stock location
   - `#<Benchmark::Tms:0x007f977a707e98 @label="", @real=20.75240011999267, @cstime=0.0, @cutime=0.0, @stime=0.23999999999999488, @utime=19.879999999999995, @total=20.11999999999999>`
- With 2000 quantity, 1000 on_hand, 1000 backordered, same stock location
   - `#<Benchmark::Tms:0x007f9772e83c10 @label="", @real=74.15757250699971, @cstime=0.0, @cutime=0.0, @stime=0.47000000000000597, @utime=72.74000000000001, @total=73.21000000000001>`
- With 3000 quantity, 1500 on_hand, 1500 backordered, same stock location
   - `#<Benchmark::Tms:0x007f9778f70af0 @label="", @real=155.99242078099633, @cstime=0.0, @cutime=0.0, @stime=0.740000000000002, @utime=153.77999999999997, @total=154.51999999999998>`

NEW
- With 10 quantity, 5 on_hand, 5 backordered, same stock location
   - `#<Benchmark::Tms:0x007fd243b61698 @label="", @real=0.17767749101039954, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.16000000000000014, @total=0.16000000000000014>`
- With 50 quantity, 25 on_hand, 25 backordered, same stock location
   - `#<Benchmark::Tms:0x007fd24bb1d1e8 @label="", @real=0.40282022999599576, @cstime=0.0, @cutime=0.0, @stime=0.019999999999999574, @utime=0.3500000000000014, @total=0.370000000000001>`
- With 100 quantity, 50 on_hand, 50 backordered, same stock location
   - `#<Benchmark::Tms:0x007fd234351d68 @label="", @real=0.6378626839723438, @cstime=0.0, @cutime=0.0, @stime=0.00999999999999801, @utime=0.5599999999999952, @total=0.5699999999999932>`
- With 500 quantity, 250 on_hand, 250 backordered, same stock location
   - `#<Benchmark::Tms:0x007fd23317c930 @label="", @real=2.8704579990007915, @cstime=0.0, @cutime=0.0, @stime=0.10000000000000142, @utime=2.450000000000003, @total=2.5500000000000043>`
- With 1000 quantity, 1 on_hand, 999 backordered, same stock location
   - `#<Benchmark::Tms:0x007fd24bfe7818 @label="", @real=5.560582629026612, @cstime=0.0, @cutime=0.0, @stime=0.17999999999999972, @utime=4.77000000000001, @total=4.95000000000001>`
- With 2000 quantity, 1000 on_hand, 1000 backordered, same stock location
   - `#<Benchmark::Tms:0x007fd24b9cf1b0 @label="", @real=9.95720759301912, @cstime=0.0, @cutime=0.0, @stime=0.3299999999999983, @utime=8.509999999999991, @total=8.83999999999999>`
- With 3000 quantity, 1500 on_hand, 1500 backordered, same stock location
   - `#<Benchmark::Tms:0x007fd24f3ebad8 @label="", @real=14.934469492000062, @cstime=0.0, @cutime=0.0, @stime=0.490000000000002, @utime=12.840000000000003, @total=13.330000000000005>`
- With 4000 quantity, 2000 on_hand, 2000 backordered, same stock location
   - `#<Benchmark::Tms:0x007fd233710bf8 @label="", @real=19.332995014003245, @cstime=0.0, @cutime=0.0, @stime=0.5700000000000003, @utime=16.689999999999998, @total=17.259999999999998>`
- With 5000 quantity, 2500 on_hand, 2500 backordered, same stock location
   - `#<Benchmark::Tms:0x007fd24b9a0d60 @label="", @real=24.674380858021323, @cstime=0.0, @cutime=0.0, @stime=0.8200000000000003, @utime=21.080000000000013, @total=21.900000000000013>`

File: core/spec/models/spree/stock/inventory_unit_builder_spec.rb [hunk 2]
```diff
@@ -13,14 +13,14 @@ module Stock
...
           units = subject.units
           expect(units.count).to eq 3
-          expect(units.first.line_item).to eq line_item_1
-          expect(units.first.variant).to eq line_item_1.variant
+          expect(units.first.line_item_id).to eq line_item_1.id
+          expect(units.first.variant_id).to eq line_item_1.variant_id
 
-          expect(units[1].line_item).to eq line_item_2
-          expect(units[1].variant).to eq line_item_2.variant
+          expect(units[1].line_item_id).to eq line_item_2.id
```
Comment (damianlegawiec) [inline core/spec/models/spree/stock/inventory_unit_builder_spec.rb]: @priyank-gupta why are we testing it like this? Aren't testing equality against records wouldn't be better than just ids? Id can be set but record can be missing.

Comment (priyank-gupta) [inline core/spec/models/spree/stock/inventory_unit_builder_spec.rb]: @damianlegawiec Yup, agreed.

File: core/app/models/spree/stock/prioritizer.rb [hunk 3]
```diff
@@ -17,22 +17,32 @@ def prioritized_packages
...
+      def adjust_packages
+        packages.each do |package|
+          package.contents.each do |item|
+            adjuster = find_adjuster(item)
+
+            if adjuster.nil?
+              adjuster = build_adjuster(item, package)
+
+            elsif item.state == :on_hand && adjuster.status == :backordered
+              # Adjust old package content if it was backordered and reassign package and state
```
Comment (Mafi88) [inline core/app/models/spree/stock/prioritizer.rb]: Could you please elaborate this comment? Why package is replaced by adjuster package, and adjuster package by package?

Comment (priyank-gupta) [inline core/app/models/spree/stock/prioritizer.rb]: @Mafi88 If item state is `on_hand` and adjuster status found for that item is `backordered` then new package (with `on_hand` item) should be assigned to adjuster and status should be changed to `on_hand` whereas old package then readjusted with adjuster.

**Comment by Mafi88** [comment]
I think there should albo be spec testing that inventory units not associated to any shipment are indeed destroyed.

**Comment by priyank-gupta** [comment]
@Mafi88 Added specs for the same.

**Comment by Mafi88** [comment]
Thanks @priyank-gupta!

**Comment by cjcruz** [comment]
Hi @priyank-gupta do you have the same fix to 3.0 version?

**Comment by priyank-gupta** [comment]
@cjcruz No. But will create a PR for 3.0.

**Comment by cjcruz** [comment]
@priyank-gupta i checked your code, i am using spree without tracking inventory so default_package method from packer.rb is creating unnecessary packages from stock items that can't relation with the current variant. so i think you should move this line https://github.com/spree/spree/blob/master/core/app/models/spree/stock/packer.rb#L29  before 'if block' starts.

**Comment by priyank-gupta** [comment]
@cjcruz Can you please submit a PR for this against master ?
I will fix this in PR's for 3-1-stable and 3-0-stable.

**Comment by priyank-gupta** [comment]
@cjcruz Created PR for 3-0-stable and 3-1-stable with fix.

**Comment by priyank-gupta** [comment]
@cjcruz Please hold that thought there. I think the change is not required as stock item check is not required for `track_inventory: false`. I will revisit the code again and will update on this.

**Pull request merged**